libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/WitnessUniverse.hs

```diff
@@ -0,0 +1,915 @@
+--   3) When (HasWitness hashtype era) holds, the WitBlock can be computed from only [ProofType hashtype era]
+--      using 'getTypeHashed'. This makes Gen and CBOR instances, especially easy. We compute only with
+--      [ProofType hashtype era] and then reconstruct the rest
+--   4) WitBlock is a Monoid, so we can combine them easily
+--   5) We can easily make (Gen (WitBlock t era)), so we can make them for testing.
+data WitBlock t era where
+  WitBlock :: (Era era, HasWitness t era) => Set t -> Map t (ProofType t era) -> WitBlock t era
+
+instance ToExpr t => ToExpr (WitBlock t era) where
+  toExpr (WitBlock x y) = Tree.App "WitBlock" [toExpr x, toExpr y]
+
+wbHash :: WitBlock t era -> Set t
+wbHash (WitBlock x _) = x
+wbMap :: WitBlock t era -> Map t (ProofType t era)
+wbMap (WitBlock _ y) = y
+
+-- | when we print a WitBlock, we are only interested in the hashes, not the witnesses
+instance PrettyA (WitBlock t era) where
+  prettyA (WitBlock hashset _) = ppSet (prettyHash @t @era) hashset
+
+instance Show (WitBlock t era) where
+  show x = show (prettyA x)
+
+instance NFData (WitBlock t era) where
+  rnf (WitBlock x y) = deepseq (rnf x) (deepseq (rnf y) ())
+
+instance Eq (WitBlock t era) where
+  (WitBlock x y) == (WitBlock a b) = x == a && y == b
+
+-- ==========================================================================
+-- (HasWitness t era) Instances for several different types of hash and hashed types
+
+-- A short descriptive name for a long complicated thing
+type BodyHash era = SafeHash EraIndependentTxBody
+
+-- ========
+-- KeyHash and VKey
+-- KeyPair seems to be missing a lot of instances, so we define them here
+
+instance Typeable r => EncCBOR (KeyPair r) where
+  encCBOR (KeyPair x y) = encode $ Rec KeyPair !> To x !> To y
+
+deriving instance Typeable r => Eq (KeyPair r)
+
+instance ToExpr (KeyPair r) where
+  toExpr (KeyPair x y) = Tree.App "KeyPair" [toExpr x, Tree.App (take 10 (show y)) []]
+
+instance Reflect era => HasWitness (KeyHash 'Witness) era where
+  type ProofType (KeyHash 'Witness) era = KeyPair 'Witness
+  type WitnessType (KeyHash 'Witness) era = (BodyHash era -> WitVKey 'Witness)
+  type TypeHashed (KeyHash 'Witness) era = VKey 'Witness
+  hash x = hashKey x
+  mkWitness keypair safehash = mkWitnessVKey safehash keypair
+  getTypeHashed (KeyPair x _) = x
+  prettyHash x = prettyA x
+
+-- ========
+-- ScriptHash and Scripts
+-- NewScript needed because (Script era) is a type family, and hence can't have instances
+
+newtype NewScript era = NewScript {unNew :: Script era}
+deriving newtype instance NFData (Script era) => NFData (NewScript era)
+deriving newtype instance EraScript era => Eq (NewScript era)
+deriving newtype instance EraScript era => EncCBOR (NewScript era)
+instance (EraScript era, ToExpr (NativeScript era)) => ToExpr (NewScript era) where
+  toExpr (NewScript script) = case getNativeScript script of
+    Just s -> toExpr s
+    Nothing -> Tree.App "PlutusScript" []
```

Comment:
I believe this is totally unnecessary, since all types that `Script` type family resolves to already have all of those instances.

So, why is this `NewScript` definition needed?

Reply:
One reason was to give it its own ToExpr instance. But I guess that's not too important. I got rid of it.